File: c/sorting/merge_sort.c
#include<stdio.h>

/**
 * Merge sort
 * - O(nlogn)
 * - Divide and conquer algorithm
 * - Outperforms n^2 sorts for all but very small datasets
 * - Requires additional storage for helper array
 * - Stable
**/

void merge_sort(int arr[], size_t size);
void split_merge(int arr[], int i_start, int i_end, int helper[]);
void merge(int arr[], int i_start, int i_mid, int i_end, int helper[]);
void print(int arr[], size_t size);

int main() {
     int arr[] = { 4, 2, 7, 3, 5, 6, 0, 1 };
     size_t size = sizeof(arr) / sizeof(arr[0]);
     merge_sort(arr, size);
     print(arr, size);
}

void merge_sort(int arr[], size_t size)
{
    int helper[size];
    split_merge(arr, 0, size - 1, helper);
}

void split_merge(int arr[], int i_start, int i_end, int helper[])
{
    if (i_end <= i_start)
    {
        return;     // Array size is 1, don't split any further
    }
    int i_mid = (i_end + i_start) / 2;              // Left subarray is arr[i_start..mid]
                                                    // Right subarray is arr[(mid + 1)..i_end]
    split_merge(arr, i_start, i_mid, helper);       // Sort the left subarray recursively
    split_merge(arr, i_mid + 1, i_end, helper);     // Sort the right subarry recursively
    // Merge the sorted subarrays
    merge(arr, i_start, i_mid, i_end, helper);
}

void merge(int arr[], int i_start, int i_mid, int i_end, int helper[])
{
    int i_left_half = i_start;        // Counter to track position in the left subarray
    int i_right_half = i_mid + 1;     // Counter to track position in the right subarray
    
    for (int k = i_start; k <= i_end; k++)   // k represents position in the original array
    {
        /* If there are items in the left half,
           and the right half is empty
           or the current item in the left half is <= the current item in the right half,
           add it to the helper array. */
        if (i_left_half <= i_mid && (i_right_half > i_end || arr[i_left_half] <= arr[i_right_half]))
        {
            helper[k] = arr[i_left_half];
            i_left_half++;
        }
        else
        {
            helper[k] = arr[i_right_half];
            i_right_half++;
        }
    }
    // Copy the sorted helper array to the original
    for (int k = i_start; k <= i_end; k++) {
        arr[k] = helper[k];
    }
}

void print(int arr[], size_t size)
{
    for (int i = 0; i < size; i++)
    {
        printf("%i ", arr[i]);
    }
    printf("\n");
}

File: ruby/sorting/merge_sort.rb
=begin
  * Merge sort
  * - Θ(nlogn)
  * - O(nlogn)
  * - Ω(nlogn)
  * - Divide and conquer algorithm
  * - Out-of-place: needs additional storage for helper array
  * - Stable
  * - Outperforms quicksort for large datasets
=end

def merge_sort(arr)
  return arr unless arr.length > 1

  # Recursively split the array down to single digits
  mid = arr.length / 2
  first_half = merge_sort(arr[0...mid])
  second_half = merge_sort(arr[mid..-1])

  # Merge the halves by comparing their first elements and shifting
  sorted = []
  while first_half.any? && second_half.any?
    sorted << (first_half[0] < second_half[0] ? first_half.shift : second_half.shift)
  end

  # Add the remaining element from either the first half or the second half
  sorted + first_half + second_half
end

numbers = [ 4, 2, 7, 3, 5, 6, 0, 1 ]
sorted = merge_sort(numbers)
puts sorted